diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Core/Unix/CoreUnix.cpp b/Core/Unix/CoreUnix.cpp
--- a/Core/Unix/CoreUnix.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Core/Unix/CoreUnix.cpp	2012-07-20 10:47:07.974481751 +0100
@@ -74,7 +74,15 @@
 		args.push_back ("--");
 		args.push_back (mountPoint);
 
+#ifndef TC_ANDROID
 		Process::Execute ("umount", args);
+#else
+		if( access( "/system/xbin/umount", F_OK ) != -1 ) {
+			Process::Execute ("/system/xbin/umount", args);
+                } else {
+			Process::Execute ("umount", args);
+                }
+#endif
 	}
 
 	shared_ptr <VolumeInfo> CoreUnix::DismountVolume (shared_ptr <VolumeInfo> mountedVolume, bool ignoreOpenFiles, bool syncVolumeInfo)
@@ -120,7 +128,15 @@
 		{
 			try
 			{
+#ifndef TC_ANDROID
 				Process::Execute ("umount", args);
+#else
+				if( access( "/system/xbin/umount", F_OK ) != -1 ) {
+					Process::Execute ("/system/xbin/umount", args);
+		                } else {
+					Process::Execute ("umount", args);
+		                }
+#endif
 				break;
 			}
 			catch (ExecutedProcessFailed&)
@@ -326,8 +342,12 @@
 	
 	string CoreUnix::GetTempDirectory () const
 	{
+#ifndef TC_ANDROID
 		char *envDir = getenv ("TMPDIR");
 		return envDir ? envDir : "/tmp";
+#else
+		return "/sdcard/Android/data/csh.cryptonite";
+#endif
 	}
 
 	bool CoreUnix::IsMountPointAvailable (const DirectoryPath &mountPoint) const
@@ -373,7 +393,15 @@
 		args.push_back (devicePath);
 		args.push_back (mountPoint);
 
+#ifndef TC_ANDROID
 		Process::Execute ("mount", args);
+#else
+		if( access( "/system/xbin/mount", F_OK ) != -1 ) {
+			Process::Execute ("/system/xbin/mount", args);
+                } else {
+			Process::Execute ("mount", args);
+                }
+#endif
 	}
 
 	VolumeSlotNumber CoreUnix::MountPointToSlotNumber (const DirectoryPath &mountPoint) const
@@ -533,14 +561,18 @@
 
 			try
 			{
+#ifndef TC_ANDROID
 				try
 				{
 					MountVolumeNative (volume, options, fuseMountPoint);
 				}
 				catch (NotApplicable&)
 				{
+#endif
 					MountAuxVolumeImage (fuseMountPoint, options);
+#ifndef TC_ANDROID
 				}
+#endif
 			}
 			catch (...)
 			{
@@ -553,6 +585,9 @@
 		{
 			try
 			{
+#if defined(TC_ANDROID) && defined(DEBUG)
+                                std::cerr << "Mount failure; aborting now" << std::endl;
+#endif
 				VolumeInfoList mountedVolumes = GetMountedVolumes (*options.Path);
 				if (mountedVolumes.size() > 0)
 				{
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Core/Unix/Linux/CoreLinux.cpp b/Core/Unix/Linux/CoreLinux.cpp
--- a/Core/Unix/Linux/CoreLinux.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Core/Unix/Linux/CoreLinux.cpp	2012-07-20 10:47:07.974481751 +0100
@@ -22,6 +22,110 @@
 #include "Driver/Fuse/FuseService.h"
 #include "Core/Unix/CoreServiceProxy.h"
 
+#ifdef TC_ANDROID
+/*
+ * Copyright (C) 2000-2006 Erik Andersen <andersen@uclibc.org>
+ *
+ * Licensed under the LGPL v2.1, see the file COPYING.LIB in this tarball.
+ */
+
+/* Reentrant version of getmntent.  */
+    struct mntent *getmntent_r (FILE *filep,
+                                struct mntent *mnt, char *buff, int bufsize)
+{
+    char *cp, *ptrptr;
+    const char *sep = " \t\n";
+
+    if (!filep || !mnt || !buff)
+        return NULL;
+
+    /* Loop on the file, skipping comment lines. - FvK 03/07/93 */
+    while ((cp = fgets(buff, bufsize, filep)) != NULL) {
+        if (buff[0] == '#' || buff[0] == '\n')
+            continue;
+        break;
+    }
+
+    /* At the EOF, the buffer should be unchanged. We should
+     * check the return value from fgets ().
+     */
+    if (cp == NULL)
+        return NULL;
+
+    ptrptr = 0;
+    mnt->mnt_fsname = strtok_r(buff, sep, &ptrptr);
+    if (mnt->mnt_fsname == NULL)
+        return NULL;
+
+    mnt->mnt_dir = strtok_r(NULL, sep, &ptrptr);
+    if (mnt->mnt_dir == NULL)
+        return NULL;
+
+    mnt->mnt_type = strtok_r(NULL, sep, &ptrptr);
+    if (mnt->mnt_type == NULL)
+        return NULL;
+
+    mnt->mnt_opts = strtok_r(NULL, sep, &ptrptr);
+    if (mnt->mnt_opts == NULL)
+        mnt->mnt_opts = (char*)"";
+
+    cp = strtok_r(NULL, sep, &ptrptr);
+    mnt->mnt_freq = (cp != NULL) ? atoi(cp) : 0;
+
+    cp = strtok_r(NULL, sep, &ptrptr);
+    mnt->mnt_passno = (cp != NULL) ? atoi(cp) : 0;
+
+    return mnt;
+}
+
+struct mntent *getmntent(FILE * filep)
+{
+    struct mntent *tmp;
+    static char *buff = NULL;
+    static struct mntent mnt;
+
+    if (!buff) {
+        buff = (char*)malloc(BUFSIZ);
+        if (!buff)
+            abort();
+    }
+
+    tmp = getmntent_r(filep, &mnt, buff, BUFSIZ);
+
+    return(tmp);
+}
+
+int addmntent(FILE * filep, const struct mntent *mnt)
+{
+    if (fseek(filep, 0, SEEK_END) < 0)
+        return 1;
+
+    if (fprintf (filep, "%s %s %s %s %d %d\n", mnt->mnt_fsname, mnt->mnt_dir,
+                 mnt->mnt_type, mnt->mnt_opts, mnt->mnt_freq, mnt->mnt_passno) < 1)
+        return 1;
+
+    return 0;
+}
+
+char *hasmntopt(const struct mntent *mnt, const char *opt)
+{
+    return strstr(mnt->mnt_opts, opt);
+}
+
+FILE *setmntent(const char *name, const char *mode)
+{
+    return fopen(name, mode);
+}
+
+int endmntent(FILE * filep)
+{
+    if (filep != NULL)
+        fclose(filep);
+    return 1;
+}
+
+#endif
+
 namespace TrueCrypt
 {
 	CoreLinux::CoreLinux ()
@@ -35,9 +139,13 @@
 	DevicePath CoreLinux::AttachFileToLoopDevice (const FilePath &filePath, bool readOnly) const
 	{
 		list <string> loopPaths;
+#ifndef TC_ANDROID
 		loopPaths.push_back ("/dev/loop");
 		loopPaths.push_back ("/dev/loop/");
 		loopPaths.push_back ("/dev/.static/dev/loop");
+#else
+		loopPaths.push_back ("/dev/block/loop");
+#endif
 
 		for (int devIndex = 0; devIndex < 256; devIndex++)
 		{
@@ -222,6 +330,7 @@
 		MountedFilesystemList mountedFilesystems;
 		DevicePath realDevicePath = devicePath;
 
+#ifndef TC_ANDROID
 		if (!devicePath.IsEmpty())
 		{
 			char *resolvedPath = realpath (string (devicePath).c_str(), NULL);
@@ -231,6 +340,7 @@
 				free (resolvedPath);
 			}
 		}
+#endif
 
 		FILE *mtab = fopen ("/etc/mtab", "r");
 
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Driver/Fuse/Driver.make b/Driver/Fuse/Driver.make
--- a/Driver/Fuse/Driver.make	2012-02-07 10:36:48.000000000 +0000
+++ b/Driver/Fuse/Driver.make	2012-07-20 10:47:07.974481751 +0100
@@ -7,10 +7,10 @@
 #
 
 NAME := Driver
-
+FUSEDIR := $(HOME)/cryptonite/fuse
 OBJS :=
 OBJS += FuseService.o
 
-CXXFLAGS += $(shell pkg-config fuse --cflags)
+CXXFLAGS += -I$(FUSEDIR)/jni/include
 
 include $(BUILD_INC)/Makefile.inc
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Driver/Fuse/FuseService.cpp b/Driver/Fuse/FuseService.cpp
--- a/Driver/Fuse/FuseService.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Driver/Fuse/FuseService.cpp	2012-07-24 18:45:54.890668674 +0100
@@ -179,8 +179,11 @@
 		}
 		return -ENOENT;
 	}
-
+#ifndef TC_ANDROID
 	static int fuse_service_read (const char *path, char *buf, size_t size, off_t offset, struct fuse_file_info *fi)
+#else
+	static int fuse_service_read (const char *path, char *buf, size_t size, loff_t offset, struct fuse_file_info *fi)
+#endif
 	{
 		try
 		{
@@ -230,7 +233,11 @@
 				shared_ptr <Buffer> infoBuf = FuseService::GetVolumeInfo();
 				BufferPtr outBuf ((byte *)buf, size);
 
+#ifndef TC_ANDROID
 				if (offset >= (off_t) infoBuf->Size())
+#else
+				if (offset >= (loff_t) infoBuf->Size())
+#endif
 					return 0;
 
 				if (offset + size > infoBuf->Size())
@@ -248,7 +255,11 @@
 		return -ENOENT;
 	}
 
+#ifndef TC_ANDROID
 	static int fuse_service_readdir (const char *path, void *buf, fuse_fill_dir_t filler, off_t offset, struct fuse_file_info *fi)
+#else
+	static int fuse_service_readdir (const char *path, void *buf, fuse_fill_dir_t filler, loff_t offset, struct fuse_file_info *fi)
+#endif
 	{
 		try
 		{
@@ -271,7 +282,11 @@
 		return 0;
 	}
 
+#ifndef TC_ANDROID
 	static int fuse_service_write (const char *path, const char *buf, size_t size, off_t offset, struct fuse_file_info *fi)
+#else
+	static int fuse_service_write (const char *path, const char *buf, size_t size, loff_t offset, struct fuse_file_info *fi)
+#endif
 	{
 		try
 		{
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Main/CommandLineInterface.cpp b/Main/CommandLineInterface.cpp
--- a/Main/CommandLineInterface.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Main/CommandLineInterface.cpp	2012-07-20 10:47:07.974481751 +0100
@@ -479,7 +479,11 @@
 			ArgVolumes = GetMountedVolumes (parser.GetParamCount() > 0 ? parser.GetParam (0) : wxString());
 
 		if (ArgCommand == CommandId::None && Application::GetUserInterfaceType() == UserInterfaceType::Text)
+#ifndef TC_ANDROID                    
 			parser.Usage();
+#else
+			ArgCommand = CommandId::Help;
+#endif
 	}
 
 	CommandLineInterface::~CommandLineInterface ()
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Main/Main.make b/Main/Main.make
--- a/Main/Main.make	2012-02-07 10:36:48.000000000 +0000
+++ b/Main/Main.make	2012-07-20 10:47:07.974481751 +0100
@@ -96,7 +96,7 @@
 
 #------ FUSE configuration ------
 
-FUSE_LIBS = $(shell pkg-config fuse --libs)
+FUSE_LIBS = -L$(HOME)/cryptonite/fuse/obj/local/armeabi -lfuse
 
 
 #------ Executable ------
@@ -105,11 +105,11 @@
 
 $(APPNAME): $(LIBS) $(OBJS)
 	@echo Linking $@
-	$(CXX) -o $(APPNAME) $(LFLAGS) $(OBJS) $(LIBS) $(FUSE_LIBS) $(WX_LIBS)
+	$(CXX) -o $(APPNAME) $(LFLAGS) $(OBJS) $(LIBS) $(FUSE_LIBS) $(WX_LIBS) $(LIBSTDCXXLIB)
 
 ifeq "$(TC_BUILD_CONFIG)" "Release"
 ifndef NOSTRIP
-	strip $(APPNAME)
+	$(HOME)/android-toolchain/bin/arm-linux-androideabi-strip $(APPNAME)
 endif
 
 ifndef NOTEST
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Main/TextUserInterface.cpp b/Main/TextUserInterface.cpp
--- a/Main/TextUserInterface.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Main/TextUserInterface.cpp	2012-07-20 10:47:07.974481751 +0100
@@ -1247,7 +1247,11 @@
 		if (!TextInputStream.get() || feof (stdin) || ferror (stdin))
 			throw UserAbort (SRC_POS);
 
+#ifndef TC_ANDROID
 		line = TextInputStream->ReadLine();
+#else
+		line = wstring(TextInputStream->ReadLine());
+#endif
 
 		if (ferror (stdin) || (line.empty() && feof (stdin)))
 			throw UserAbort (SRC_POS);
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Main/UserInterface.cpp b/Main/UserInterface.cpp
--- a/Main/UserInterface.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Main/UserInterface.cpp	2012-07-20 10:47:07.974481751 +0100
@@ -1400,7 +1400,12 @@
 		EncryptionTest::TestAll();
 
 		// StringFormatter
+#ifndef TC_ANDROID
 		if (StringFormatter (L"{9} {8} {7} {6} {5} {4} {3} {2} {1} {0} {{0}}", "1", L"2", '3', L'4', 5, 6, 7, 8, 9, 10) != L"10 9 8 7 6 5 4 3 2 1 {0}")
+#else
+                wxString testString = StringFormatter (L"{9} {8} {7} {6} {5} {4} {3} {2} {1} {0} {{0}}", "1", L"2", '3', L'4', 5, 6, 7, 8, 9, 10);
+		if (testString != L"10 9 8 7 6 5 4 3 2 1 {0}")
+#endif
 			throw TestFailed (SRC_POS);
 		try
 		{
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Platform/StringConverter.cpp b/Platform/StringConverter.cpp
--- a/Platform/StringConverter.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Platform/StringConverter.cpp	2012-07-20 10:47:07.978481765 +0100
@@ -213,6 +213,7 @@
 	{
 		try
 		{
+#ifndef TC_ANDROID
 			mbstate_t mbState;
 			Memory::Zero (&mbState, sizeof (mbState));
 			const wchar_t *src = wstr.c_str();
@@ -230,6 +231,11 @@
 			str.clear();
 			str.insert (0, &buf.front(), size);
 			Memory::Erase (&buf.front(), buf.size());
+#else
+                        str.clear();
+                        str.assign(wstr.begin(), wstr.end());
+                        str = string(str.c_str());
+#endif
 		}
 		catch (...)
 		{
@@ -298,6 +304,7 @@
 	{
 		try
 		{
+#ifndef TC_ANDROID                    
 			mbstate_t mbState;
 			Memory::Zero (&mbState, sizeof (mbState));
 			const char *src = str.c_str();
@@ -315,6 +322,11 @@
 			wstring s;
 			s.insert (s.begin(), buf.begin(), buf.begin() + size);
 			return s;
+#else
+                        wstring s;
+                        s.insert(s.begin(), str.begin(), str.end());
+                        return s;
+#endif
 		}
 		catch (...)
 		{
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Platform/Unix/File.cpp b/Platform/Unix/File.cpp
--- a/Platform/Unix/File.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Platform/Unix/File.cpp	2012-07-20 10:47:07.978481765 +0100
@@ -15,6 +15,11 @@
 #include <sys/mount.h>
 #endif
 
+#ifdef TC_ANDROID
+#include <sys/ioctl.h>
+#include <linux/fs.h>
+#endif
+
 #ifdef TC_BSD
 #include <sys/disk.h>
 #endif
@@ -174,17 +179,25 @@
 #	endif
 		}
 #endif
+#ifdef TC_ANDROID
+		loff_t current = lseek64 (FileHandle, 0, SEEK_CUR);
+#else
 		off_t current = lseek (FileHandle, 0, SEEK_CUR);
+#endif
 		throw_sys_sub_if (current == -1, wstring (Path));
 		SeekEnd (0);
+#ifdef TC_ANDROID
+		uint64 length = lseek64 (FileHandle, 0, SEEK_CUR);
+#else
 		uint64 length = lseek (FileHandle, 0, SEEK_CUR);
+#endif
 		SeekAt (current);
 		return length;
 	}
 
 	void File::Open (const FilePath &path, FileOpenMode mode, FileShareMode shareMode, FileOpenFlags flags)
 	{
-#ifdef TC_LINUX
+#if defined(TC_LINUX) && !defined(TC_ANDROID)
 		int sysFlags = O_LARGEFILE;
 #else
 		int sysFlags = 0;
@@ -224,7 +237,12 @@
 			ModTime = statData.st_mtime;
 		}
 
+#ifndef TC_ANDROID
 		FileHandle = open (string (path).c_str(), sysFlags, S_IRUSR | S_IWUSR);
+#else
+		string ansii_path = StringConverter::ToSingle(wstring(path));
+		FileHandle = open (ansii_path.c_str(), sysFlags, S_IRUSR | S_IWUSR);
+#endif
 		throw_sys_sub_if (FileHandle == -1, wstring (path));
 
 #if 0 // File locking is disabled to avoid remote filesystem locking issues
@@ -298,7 +316,15 @@
 #ifdef TC_TRACE_FILE_OPERATIONS
 		TraceFileOperation (FileHandle, Path, false, buffer.Size(), position);
 #endif
+#ifdef TC_ANDROID
+		loff_t current = lseek64 (FileHandle, 0, SEEK_CUR);
+		throw_sys_sub_if (current == -1, string (Path));
+		SeekAt (position);
+		ssize_t bytesRead = read (FileHandle, buffer, buffer.Size());
+		SeekAt (current);
+#else
 		ssize_t bytesRead = pread (FileHandle, buffer, buffer.Size(), position);
+#endif
 		throw_sys_sub_if (bytesRead == -1, wstring (Path));
 
 		return bytesRead;
@@ -307,7 +333,11 @@
 	void File::SeekAt (uint64 position) const
 	{
 		if_debug (ValidateState());
+#ifdef TC_ANDROID
+		throw_sys_sub_if (lseek64 (FileHandle, position, SEEK_SET) == -1, wstring (Path));
+#else
 		throw_sys_sub_if (lseek (FileHandle, position, SEEK_SET) == -1, wstring (Path));
+#endif
 	}
 
 	void File::SeekEnd (int offset) const
@@ -323,7 +353,11 @@
 		}
 #endif
 
+#ifdef TC_ANDROID
+		throw_sys_sub_if (lseek64 (FileHandle, offset, SEEK_END) == -1, wstring (Path));
+#else
 		throw_sys_sub_if (lseek (FileHandle, offset, SEEK_END) == -1, wstring (Path));
+#endif
 	}
 
 	void File::Write (const ConstBufferPtr &buffer) const
@@ -343,6 +377,15 @@
 #ifdef TC_TRACE_FILE_OPERATIONS
 		TraceFileOperation (FileHandle, Path, true, buffer.Size(), position);
 #endif
+#ifdef TC_ANDROID
+		loff_t current = lseek64 (FileHandle, 0, SEEK_CUR);
+		throw_sys_sub_if (current == -1, string (Path));
+		SeekAt (position);
+		ssize_t bytesWritten = write (FileHandle, buffer, buffer.Size());
+		SeekAt (current);
+		throw_sys_sub_if (bytesWritten != (ssize_t) buffer.Size(), string (Path));
+#else
 		throw_sys_sub_if (pwrite (FileHandle, buffer, buffer.Size(), position) != (ssize_t) buffer.Size(), wstring (Path));
+#endif
 	}
 }
diff -x '*.in' -x '*.o' -x '*.lo' -x '*.a' -x '*.xml.h' -x '*.1' -x 'bld*' -x '*~' -x '*.d' -ur a/Platform/Unix/Process.cpp b/Platform/Unix/Process.cpp
--- a/Platform/Unix/Process.cpp	2012-02-07 10:36:48.000000000 +0000
+++ b/Platform/Unix/Process.cpp	2012-07-20 10:47:07.978481765 +0100
@@ -189,6 +189,9 @@
 			if (!errOutput.empty())
 				strErrOutput.insert (strErrOutput.begin(), errOutput.begin(), errOutput.end());
 
+#if defined(TC_ANDROID) && defined(DEBUG)
+                        std::cout << strErrOutput << std::endl;
+#endif
 			throw ExecutedProcessFailed (SRC_POS, processName, exitCode, strErrOutput);
 		}
 
