diff -ur ./libs/serialization/src/extended_type_info_typeid.cpp ./libs/serialization/src/extended_type_info_typeid.cpp
--- ./libs/serialization/src/extended_type_info_typeid.cpp	2011-01-19 18:31:02.000000000 +0000
+++ ./libs/serialization/src/extended_type_info_typeid.cpp	2012-01-09 12:27:08.000000000 +0000
@@ -51,9 +51,13 @@
     // shortcut for common case
     if(this == & rhs)
         return false;
+#ifdef ANDROID
+    return 0 != (m_ti->name() < (*(static_cast<const extended_type_info_typeid_0 &>(rhs).m_ti)).name());
+#else
     return 0 != m_ti->before(
         *(static_cast<const extended_type_info_typeid_0 &>(rhs).m_ti)
-    );
+                             );
+#endif
 }
 
 BOOST_SERIALIZATION_DECL(bool) 
